datafusion/core/src/execution/session_state.rs

```diff
@@ -976,6 +837,482 @@ impl SessionState {
     }
 }
 
+/// A builder to be used for building [`SessionState`]'s. Defaults will be used for all values
+/// unless explicitly provided. Note that there is no `Default` or `new()` for SessionState,
+/// to avoid accidentally running queries or other operations without passing through
+/// the [`SessionConfig`] or [`RuntimeEnv`].
+pub struct SessionStateBuilder {
+    state: SessionState,
+    use_defaults: bool,
+}
+
+impl SessionStateBuilder {
+    /// Returns new [`SessionStateBuilder`] using the provided
+    /// [`SessionConfig`] and [`RuntimeEnv`].
+    pub fn new_with_config_rt(
```

Comment:
Another way we could do this would be simply

```rust
let builder = SessionStateBuilder::new()
  .with_config(config)
  .with_runtime_env(runtime_env);
```

I think the original concern with an API like this is that this would likely create a new `SessionConfig::default()` just to replace it with `config` 

However, I think we could optimize this eventually (e.g. by deferring the creation of `SessionState` until calling `build()` for example) and if we are going to change the API anyways I think we should consider doing the "better" thing

Reply:
I had that originally however I decided to switch just to mirror the existing behavior. It's a simple change, just would need to document that .build() will fail if config and runtime_env are not provided (assuming that we don't just use defaults there)